### PR TITLE
Fix sample size inflation in R2Score Adjusted R2 calculation

### PR DESCRIPTION
**Pull request by buildwithsuhana**
## Description
This PR fixes a mathematical bug in the R2Score metric's Adjusted $R^2$ calculation where the sample size was being incorrectly inflated in multi-output regression scenarios.

### The Issue
Previously, num_samples was updated using ops.size(y_true). In multi-output regression (where y_true has shape (batch_size, num_outputs)), ops.size returns the total number of elements ($N \times K$).

The Adjusted $R^2$ formula, $1 - (1 - R^2) \frac{n-1}{n-p-1}$, requires $n$ to be the number of observations (samples). By using $N \times K$, the metric was artificially inflating the sample size, which:
   1. Deflated the penalty term for independent regressors, leading to overly optimistic scores.
   2. Bypassed degree-of-freedom safety checks (e.g., it would calculate a result even if $N \le p+1$, as long as $N \times K > p+1$).
   3. Ignored sample_weight in the sample size count, leading to further divergence in weighted scenarios.

### The Fix
The update logic for num_samples now uses ops.mean(ops.sum(sample_weight, axis=0)). This correctly recovers the number of observations $N$ for both single-output and multi-output regression and correctly accounts for sample_weight by providing the effective sample size.

Changes
   - Core Logic: Modified keras/src/metrics/regression_metrics.py to correctly update num_samples.
   - Tests: Updated keras/src/metrics/regression_metrics_test.py with corrected expectations for multi-output Adjusted $R^2$ tests. Added a new test case to verify Adjusted $R^2$ accuracy in multi-output scenarios with sufficient sample size.

Bug reproduction script: https://colab.research.google.com/drive/1vJLiamTJ4x8DGhODfd_VgQPzrV5ajCLm?usp=sharing
## Contributor Agreement

**Please review our [AI-Assisted Contribution Policy](../CONTRIBUTING.md#ai-assisted-contribution-policy) and check all boxes below before submitting your PR for review:**

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

**Note:** Failing to adhere to this agreement may result in your future PRs no longer being reviewed.

### CODE REVIEW

File: keras/src/metrics/regression_metrics.py [hunk 1]
```diff
@@ -522,7 +522,21 @@ def update_state(self, y_true, y_pred, sample_weight=None):
...
+                    ops.sum(
+                        ops.cast(
+                            ops.not_equal(sample_weight, 0.0),
+                            dtype=self.num_samples.dtype,
+                        ),
+                        axis=0,
+                    )
+                ),
+                dtype=self.num_samples.dtype,
+            )
```
Comment (hertschuh) [inline keras/src/metrics/regression_metrics.py]: This is very unreadable. Please create intermediary values so that things fit on fewer lines with one indentation at most.

But also, I don't agree with the formula.

- the mean doesn't make sense to me, it will create fractional values, but `num_sample` should be an integral value, the number of samples
- `sample_weight` should be 1D or maybe [batch, 1], so yes, you do want to count the non-zero values, but a `sum` across all axes should be enough
- then you wouldn't need 2 casts to the same dtype

**Review by hertschuh — CHANGES_REQUESTED** [review]
Thanks!

**Comment by hertschuh** [comment]
Oh, I forgot to ask, in your colab, can you compare with sklearn to make sure the formula is right?

File: keras/src/metrics/regression_metrics.py (hunk 1 above)
Comment (buildwithsuhana) [reply keras/src/metrics/regression_metrics.py]: I've refactored this to correctly count unique samples using ops.any, ensuring num_samples stays integral and matches the adjusted R2 formula.

**Comment by buildwithsuhana** [comment]
> Oh, I forgot to ask, in your colab, can you compare with sklearn to make sure the formula is right?

Done, I have added the test in the colab to show the mismatch.

**Review by hertschuh — APPROVED** [review]
Thanks for the fix!
